### ISSUE TITLE
ZeroTier through BMX6 routing

### ISSUE THREAD

**Issue by nickbash11** [zerotier/ZeroTierOne#1131]
In my case I need to use ZeroTier on my MESH router with BMX6. That means a router has no any default gateway in its routing table, only subnets, for example:

```
# route -n
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
10.12.11.1      0.0.0.0         255.255.255.255 UH    1024   0        0 bmxOut_qMp-dd4
10.55.1.0       0.0.0.0         255.255.255.224 U     1024   0        0 bmxC4tmain
10.40.0.1       0.0.0.0         255.255.255.255 UH    1024   0        0 bmxOut_qMp-1ee
10.51.1.0       0.0.0.0         255.255.255.224 U     1024   0        0 bmxC4tmain
10.44.1.0       0.0.0.0         255.255.255.224 U     1024   0        0 bmxC4tmain
10.70.1.0       0.0.0.0         255.255.255.224 U     0      0        0 br-lan
169.254.6.0     0.0.0.0         255.255.255.248 U     0      0        0 wlan0
192.168.196.0   0.0.0.0         255.255.255.0   U     0      0        0 ztbtoswaiv
```

I've noticed that ZeroTier does not work while the device has no default route. How can I be in this situation?

```
# zerotier-cli info
200 info a704dad35e 1.4.6 OFFLINE
```

**Comment by glimberg** [comment]
ZeroTier needs a way out to the internet to contact root servers and network controllers.  Without root servers, it can't find other nodes.  Without network controllers, network configuration cannot be received.

**Comment by nickbash11** [comment]
I understand that, but the MESH router has an internet access over BMX6. Issue is that the ZeroTier cannot access it without DEFAULT GATEWAY in routing table.

Ok, I'll try to check it with tcpdump or something else.

**Comment by nickbash11** [comment]
I captured packets from the mesh node, might be somebody will tell me what wrong is...

[zerotier1.pcap.txt](https://github.com/zerotier/ZeroTierOne/files/4127960/zerotier1.pcap.txt)

**Comment by paweljacewicz** [comment]
Hi @nickbash11 

I did some experiments on a "LAN without default GW" a long time ago and from what I remember I used the moon functionality to help solve some issues. Can you try an experiment and:
1. Deploy ZT on an additional router/mesh node that has the default gateway set in the routing table (and is able to reach the internet via it).
2. Create the moon file with configuration of the ZT instance from point 1. Especially put the locally reachable IP address there, so nodes on the LAN can connect to it without default GW. Deploy the moon on the node, restart ZT. Some guide: https://www.zerotier.com/manual/#4_4 
3. Deploy the moon file on the mesh router that is currently having the issues, restart ZT.

Let us know if that helps.

**Comment by paweljacewicz** [comment]
OK... I did some testing and on LAN without default gateway ZT doesn't want to work, even with moon server deployed as an endpoint with Internet connectivity (two interfaces, one LAN, one NAT to internet) which is reachable by other ZT nodes on LAN (shows in peer list with its LAN IP address). I assumed the moon would act as a forwarder of a kind, based on info about how ROOT servers behave, on https://www.zerotier.com/manual/#2_1_1.

@adamierymenko Would find a moment and comment on the issue, please? Maybe I'm missing something, but the manual on 2.1.1 says that ROOT servers forward packets. Do MOON servers also do that?

Thanks in advance!

**Comment by nickbash11** [comment]
Thanks for reply!

I've tried to get a moon. Accordingly [4_4](https://www.zerotier.com/manual/#4_4):
1. I generated a moon file on a device which I suppose to use as a moon (It has a direct connection to the internet.): zerotier-idtool initmoon <itself identity.public> > moon.json . 
2. added to it an ip address itself (an address of a node where I generated the moon.json)
3. did zerotier-idtool genmoon moon.json
4. sent the obtained moon file (*.moon) to the directory /var/lib/zerotier-one/moons.d/ on ANOTHER regular zerotier node which needs it.

Did I right? Correct me.

So, the last one node did not see anything moons in peers. But, I've tried to place moon file on my work laptop (which has default route), and it sees something MOON as a relay without PATH. I'll attach a screen from my laptop.
![Screenshot_2020-02-03_15-31-44](https://user-images.githubusercontent.com/3481744/73653586-8fd17c00-469a-11ea-94b6-4bed74dbb9a1.png)

5dcf664936 is a node which I suppose to use a moon and generated moon file on it.

**Comment by nickbash11** [comment]
I got one mesh node connected through my laptop as a moon, but anyway it does not obtain ip configuration and I do not see any state connection in **netstat** with my moon, only SYN requests to 204.80.128.1:443

So, another one problem what I have discovered is the BMX6-specifically. No one peer can reach to each other through 9993 port, it rejects all for some reasons, e.g. port 9993 on a MESH router rejects any connection to its IP address. It explains why I see that picture on a screenshot presented above.

BMX6/7 has many interfaces with identical IP address also includes br-lan, might be a problem. It creates a new point-to-point tun device for each communication with another MESH node. I think maintainers have to deploy something MESH protocol and check how the zerotier interacts

**Comment by paweljacewicz** [comment]
Hi! Can you check some settings so I have a clear picture?

> 1. I generated a moon file on a device which I suppose to use as a moon (It has a direct connection to the internet.): **zerotier-idtool initmoon (itself identity.public) > moon.json** .
> 2. added to it an ip address itself (an address of a node where I generated the moon.json)

Did you add in the moon json file an IP address that is reachable via the br-lan, as shown on the first post? I assume the br-lan connects your local machines with the mesh router?

> 3. did **zerotier-idtool genmoon moon.json**
> 4. sent the obtained moon file (*.moon) to the directory **/var/lib/zerotier-one/moons.d/** on **ANOTHER** regular zerotier node which needs it.

As far as I understand the manual, the generated moon file has to be put on two nodes: the node where you generated it on (the one with the internet connectivity) and on the other node that has issues (as you said you did in 4.). Both ZT need to be restarted (at least it won't hurt). You can check whether the moons are defined with the "zerotier-cli listmoons" command. And moons should also show on the peer list. 
Also, please check permissions and ownership of the /var/lib/zerotier-one/moons.d/ directory and files inside. All should belong to zerotier-one user.

**Comment by nickbash11** [comment]
> Did you add in the moon json file an IP address that is reachable via the br-lan, as shown on the first post? I assume the br-lan connects your local machines with the mesh router?

Further experiments followed me to use my laptop as a moon server because last one has no any problems. 

![Screenshot_2020-02-03_20-19-30](https://user-images.githubusercontent.com/3481744/73674897-852add00-46c2-11ea-8c4e-dd5311612148.png)
This screen have got from MESH router, where 10.60.1.29 is my laptop.

My moon.json for example:

```
{
 "id": "5ce8f9519e",
 "objtype": "world",
 "roots": [
  {
   "identity": "5ce8f9519e:deleted",
   "stableEndpoints": [ "10.60.1.29/9993" ]
  }
 ],
 "signingKey": "deleted",
 "signingKey_SECRET": "deleted",
 "updatesMustBeSignedBy": "deleted",
 "worldType": "moon"
}
```

After some experiments I found out that the moon file might be placed only on a regular zerotier node. But I placed it on both sides, just in case...

It works itself, but only within the same network subnet. For example zerotier from IP 10.12.11.1 cannot connect to moon, although ping and telnet reach this host...

**Comment by nickbash11** [comment]
As far as I understand it needs to deploy a network in a virtual MESH environment without internet access, where will be a zerotier controller, a moon and a regular node.

**Comment by paweljacewicz** [comment]
> > Did you add in the moon json file an IP address that is reachable via the br-lan, as shown on the first post? I assume the br-lan connects your local machines with the mesh router?
> 
> Further experiments followed me to use my laptop as a moon server because last one has no any problems.
> 
> ![Screenshot_2020-02-03_20-19-30](https://user-images.githubusercontent.com/3481744/73674897-852add00-46c2-11ea-8c4e-dd5311612148.png)
> This screen have got from MESH router, where 10.60.1.29 is my laptop.

So, if the moon started working for you, were you able to connect to some ZT network from your MESH router?

**Comment by nickbash11** [comment]
Yes, it obtained IP and can reach other hosts in the network. However, a moon server can be reached only either in the same subnet or through default gateway. I checked it recently.

**Comment by nickbash11** [comment]
The main question is still alive - why does zerotier reach neither moons (without subnet access) nor planet without default gateway? And why does not its ports (9993) available on external IPs of BMX6 nodes? The last one problem does moons useless for me.

**Comment by nickbash11** [comment]
I accidentally found that the BMX6 keeps default route in the table **"default"** not in **"main"**
I think it can be answer.

```
root@qMp-7bac:~# ip rule
0:	from all lookup local 
60:	from all lookup 60 
32766:	from all lookup main 
32767:	from all lookup default 
root@qMp-7bac:~# ip route show table default
default dev bmxOut_qMp-7d9 proto static metric 1024 
root@qMp-7bac:~# 
```
